File: 0023 - Merge k Sorted Lists/Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int amount = lists.size();
        int interval = 1;
        while (interval < amount) {
            for (int i = 0; i < amount - interval; i += interval * 2) {
                lists[i] = mergeTwoLists(lists[i], lists[i + interval]);
            }
            interval *= 2;
        }
        return amount > 0 ? lists[0] : NULL;
    }
    
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(0);
        ListNode* point = head;
        while (l1 && l2) {
            if (l1->val <= l2->val) {
                point->next = l1;
                l1 = l1->next;
            } else {
                point->next = l2;
                l2 = l2->next;
            }
            point = point->next;
        }
        if (!l1) {
            point->next = l2;
        } else {
            point->next = l1;
        }
        return head->next;
    }
};

File: 0023 - Merge k Sorted Lists/Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        interval = 1
        amount = len(lists)
        
        while interval < amount:
            for i in range(0, amount - interval, interval * 2):
                lists[i] = self.merge2Lists(lists[i], lists[i + interval])
            interval *= 2
            
        return lists[0] if amount > 0 else None
            
            
    def merge2Lists(self, first: ListNode, second: ListNode) -> ListNode:
        first_head = first
        second_head = second
        
        newHead = ListNode()
        head = newHead
        
        while first_head is not None and second_head is not None:
            if first_head.val < second_head.val:
                newHead.next = first_head
                first_head = first_head.next
            else:
                newHead.next = second_head
                second_head = second_head.next
            newHead = newHead.next
            
        if first_head is None:
            newHead.next = second_head
        else:
            newHead.next = first_head

